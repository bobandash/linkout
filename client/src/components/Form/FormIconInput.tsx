import { FC, useRef } from 'react';
import camelize from '../../utils/camelize';

interface FormIconInputProps {
  centered?: boolean;
  handleInputChange: (e: React.ChangeEvent<HTMLInputElement>) => void;
  name: string;
}

const FormIconInput: FC<FormIconInputProps> = ({
  centered,
  handleInputChange,
  name,
}) => {
  const nameCamelized = camelize(name);
  const imageRef = useRef<HTMLImageElement | null>(null);
  const loadFile = (e: React.ChangeEvent<HTMLInputElement>) => {
    if (imageRef.current !== null && e.target.files !== null) {
      imageRef.current.src = URL.createObjectURL(e.target.files[0]);
      imageRef.current.onload = function () {
        if (imageRef.current !== null) {
          URL.revokeObjectURL(imageRef.current.src);
        }
      };
    }
  };

  return (
    <div
      className={`relative flex aspect-square max-w-[80px] overflow-hidden rounded-full border-2 bg-white hover:cursor-pointer md:max-w-[100px] ${
        centered && 'mx-auto'
      }`}
    >
      <img
        alt="picture"
        className="flex-grow hover:cursor-pointer "
        ref={imageRef}
      />
      <input
        onChange={(e) => {
          loadFile(e);
          handleInputChange(e);
        }}
        type="file"
        className="absolute left-0 top-0 h-full w-full opacity-0 hover:cursor-pointer"
        name={nameCamelized}
        id={nameCamelized}
        accept="image/png, image/jpeg"
      />
    </div>
  );
};

export default FormIconInput;
