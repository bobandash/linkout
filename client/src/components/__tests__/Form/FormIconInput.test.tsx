import { render, screen, fireEvent } from '@testing-library/react';
import FormIconInput from '../../Form/FormIconInput';

const handleInputChangeMock = vi.fn();
global.URL.createObjectURL = vi.fn(() => 'mocked-object-url');
global.URL.revokeObjectURL = vi.fn();

function getById<T extends Element>(container: HTMLElement, id: string): T {
  const element = container.querySelector<T>(`#${id}`);
  assert(element !== null, `Unable to find an element with ID #${id}.`);
  return element;
}

describe('FormIconInput', () => {
  it('renders correctly', () => {
    render(
      <FormIconInput
        name="icon"
        centered={false}
        handleInputChange={handleInputChangeMock}
        defaultValue="example.jpg"
      />,
    );
    expect(FormIconInput).toMatchSnapshot();
  });

  it('renders a text placeholder if defaultValue is not an image', () => {
    render(
      <FormIconInput
        name="icon"
        centered={false}
        handleInputChange={handleInputChangeMock}
        defaultValue="random text"
      />,
    );
    expect(screen.getByText('RANDOM TEXT')).toBeInTheDocument();
  });

  it('renders an image if defaultValue is an image', () => {
    render(
      <FormIconInput
        name="icon"
        centered={false}
        handleInputChange={() => {}}
        defaultValue="images/example.jpg"
      />,
    );
    expect(screen.getByAltText('icon image')).toBeInTheDocument();
  });

  it('triggers creating an internal url to display and handleInputChange when input is changed', () => {
    const name = 'icon';
    const { container } = render(
      <FormIconInput
        name={name}
        centered={false}
        handleInputChange={handleInputChangeMock}
      />,
    );

    const input = getById<HTMLInputElement>(container, name);
    fireEvent.change(input, {
      target: {
        files: [new File([''], 'example.jpg', { type: 'image/jpeg' })],
      },
    });

    expect(handleInputChangeMock).toHaveBeenCalled();
    expect(global.URL.createObjectURL).toHaveBeenCalled();
  });
});
