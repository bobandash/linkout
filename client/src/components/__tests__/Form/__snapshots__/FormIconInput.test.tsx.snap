// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FormIconInput > renders 1`] = `[Function]`;

exports[`FormIconInput > renders correctly 1`] = `[Function]`;
